libs/teckit/TECkit-src/docs/teckit_compile.1
.TH TECKIT_COMPILE 1 "2015-11-14" "TECkit"
.SH NAME
teckit_compile \- manual page for command line error at --version
.SH SYNOPSIS
.B teckit_compile
[\fI\,-u\/\fR] [\fI\,-x\/\fR] [\fI\,-z\/\fR] \fI\,mapping_description \/\fR[\fI\,-o compiled_table\/\fR]
.SH OPTIONS
Required argument:
.IP
\fBsource\fR mapping description (.map) file
.PP
Optional arguments:
.TP
.BR \fB\-o\fR file
output compiled table (.tec) file ("\-" for stdout)
.TP
.BR \fB\-u\fR
read source text as UTF8 even if no BOM found
.TP
.BR \fB\-x\fR
generate XML representation rather than compiled table
.TP
.BR \fB\-z\fR
generate uncompressed table format
.SH Further Information
For further information see the document TECkit_Tools.pdf.
A description of the mapping description language used by the compiler, with examples, is found in TECkit_Language.pdf.
On a Debian system, these can be found in /usr/share/doc/teckit.


ChangeLog
2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the
	result of make install is not in itself usable.

2016-06-30  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2016 => 2017/dev.
	* tardate.ac (tex_live_tardate): 2015-05-12 => 2016-05-23.
	(Though this is not used for any release.)

2016-03-16  Karl Berry  <karl@tug.org>

	* Build: check for exitstatus.txt not having an exit status.

2016-03-06  Karl Berry  <karl@tug.org>

	* version.ac: remove /dev to prepare for 2016.

2015-06-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2015 => 2016/dev.
	* tardate.ac: Switch to 2015-06-12.

2015-05-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris (removed): Moved from here to doc/.
	* Makefile.am: Adapted.

2015-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris: New file from Mojca (without very long lines).
	* Makefile.am: Add the new file.

2015-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2015-03-07  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2015-03-07.

2014-11-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-11-24.

2014-10-21  Karl Berry  <karl@tug.org>

	* configure.ac: report CC/CXX values, with --version output
	if available.
	* Build: no need for objc any more.

2014-08-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-08-22.

2014-06-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2014 => 2015/dev.
	* tardate.ac: Switch to 2014-06-16.

2014-04-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Makefile.am: New names for README.* files.

2014-04-17  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2014-04-04  Karl Berry  <karl@tug.org>

	* Build (TL_MAKE,MAKE): set both in all cases.
	(buildenv.log): include pwd, and use .log instead of .txt.

2014-04-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-04-01.

2014-01-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-01-01.

2013-12-26  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (ACLOCAL): Ensure aclocal.m4 is up to date.

2013-10-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* doc: New directory for texlive.info & Co.
	* Makefile.am, configure.ac: Adapted.

2013-10-05  Karl Berry  <karl@tug.org>

	* Build: reorder variables and settings for clarity.
	(TL_CONF_BANNER): restore variable, default is empty string.
	(TL_POSTCONFIGURE): new variable, default true.

2013-09-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Drop ACLOCAL_AMFLAGS.
	* configure.ac: Use AC_CONFIG_MACRO_DIRS.
	Use KPSE_RECURSE_LIBS in m4/kpse-setup.m4 to recurse into all
	TeX specific libraries.

2013-09-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-09-01.

	* Makefile.am: Moved definition of cf_silent* from here ...
	* am/recurse.am: ... to here 

2013-08-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Add the make target 'texlinks'.

2013-07-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am: new subdirectory.
	* Makefile.am, configure.ac: Use am/recurse.am.

2013-06-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Added Automake options 'tar-pax' for paths
	longer than 100 chars and 'no-dist-gzip'.
	* version.ac: Switch 2013 => 2014/dev'
	* tardate.ac: Switch to 2013-06-24.

2013-03-26  Karl Berry  <karl@tug.org>

	* Build: use $MAKE rather than $TL_MAKE for the make clean
	in the unlikely event of a local Makefile.

2013-03-21  Mojca Miklavec  <mojca.miklavec.lists@gmail.com>

	* Build: Remove GNUMAKE variable (after freetype2 as the only library
	which needed it has been fixed).

2013-03-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-03-01.

2013-02-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Building for Windows does not require 'ln -s'.
	* README.config: Document '--without-ln-s'.

2013-02-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (mandir, infodir): Change texmf => texmf-dist.

2013-01-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-01-24.
	* Makefile.am: Running texlinks in texk/texlive/tl_scripts.

2013-01-18  Karl Berry  <karl@tug.org>

	* Build: remove use of time.

2012-12-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Enable silent rules.

2012-11-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-11-16.

2012-10-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-10-24.

2012-08-19  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-08-19.

2012-08-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Running texlinks in texk/texlive/linked_scripts.

2012-07-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build: Add TL_COMPILER_GFLAGS also to OBJCXXFLAGS.

2012-07-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2012 => 2013/dev.

2012-06-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-06-15.

2012-05-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Test for GNU make before creating Makefile,
	required for FreeBSD make and perhaps others.

2012-05-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README: Update the list of prerequisites.
	* README.config: Mention configure options for biber.

2012-04-12  Karl Berry  <karl@tug.org>

	* Build (TL_COMPILER_GFLAGS): new envvar, used to override -g
	for all three compilers, with Build -g.

2012-04-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-04-03.

2012-03-31  Karl Berry  <karl@tug.org>


libs/teckit/TECkit-src/source/Sample-tools/TECkit_Compile.cpp
/*
	TECkit_Compile.c
	Copyright (c) 2002-2016 SIL International.
	
	2004-03-12	updated to use v2.1 compiler
				added -u option to force UTF8 mode
	2005-03-18	added -x option to generate XML representation

	 5-May-2005		jk	added include <stdlib.h> to keep gcc happy
	24-May-2005		jk	change placement of CALLBACK for VC++6 (Ulrik)
	 1-Apr-2008		jk	extern "C" decl for errFunc; const char* progName
*/

#include "TECkit_Compiler.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#if __MWERKS__
#include "SIOUX.h"
#include "console.h"
#endif

extern "C" {
	static void CALLBACK errFunc(void* userData, const char* msg, const char* param, UInt32 line);
};

static
void
CALLBACK
errFunc(void* /* userData */, const char* msg, const char* param, UInt32 line)
{
	fprintf(stderr, "%s", msg);
	if (param != 0)
		fprintf(stderr, ": \"%s\"", param);
	if (line != 0)
		fprintf(stderr, " at line %lu", (unsigned long)line);
	fprintf(stderr, "\n");
}

int
main(int argc, char** argv)
{
	const char*	progName = argv[0];
	char*	mapFileName = 0;
	char*	tecFileName = 0;
	int		errorCount = 0;
	char	compress = 1;
	char	form = kForm_Unspecified;
	char	genXML = 0;
	
	if (*progName == 0)
		progName = "TECkit_Compile";
	
#if __MWERKS__ && __dest_os == __mac_os
	SIOUXSettings.asktosaveonclose = 0;
	argc = ccommand(&argv);
#endif

	while (--argc) {
		char*	arg = *++argv;
		if (arg[0] == '-' && strlen(arg + 1) == 1 && argc > 0) {
			switch (arg[1]) {
				case 'o':
					if (argc > 1) {
						tecFileName = *++argv;
						--argc;
					}
					else {
						fprintf(stderr, "missing filename after -o\n");
						++errorCount;
					}
					break;
				case 'z':
					compress = 0;
					break;
				case 'u':
					form = kForm_UTF8;
					break;
				case 'x':
					genXML = 1;
					break;
				default:
					fprintf(stderr, "unknown option: -%c\n", arg[1]);
					++errorCount;
					break;
			}
		}
		else if (arg[0] != '-' && mapFileName == 0)
			mapFileName = arg;
		else {
			fprintf(stderr, "command line error at %s\n", arg);
			++errorCount;
		}
	}

	if (errorCount > 0 || mapFileName == 0) {
		fprintf(stderr, "\
Usage: %s [-u] [-x] [-z] mapping_description [-o compiled_table]\n\
\n\
    Required argument:\n\
        source mapping description (.map) file\n\
\n\
    Optional arguments:\n\
        -o file     output compiled table (.tec) file (\"-\" for stdout)\n\
        -u          read source text as UTF8 even if no BOM found\n\
        -x          generate XML representation rather than compiled table\n\
        -z          generate uncompressed table format\n\
", progName);
		return 1;
	}

	if (tecFileName == 0) {
		int	x = strlen(mapFileName);
		tecFileName = (char*)malloc(x + 5);
		if (tecFileName == 0)
			return 1;	// unlikely!
		strcpy(tecFileName, mapFileName);
		if (x > 4) {
			if (tecFileName[x - 4] == '.') {
				--x; tecFileName[x] = tolower(tecFileName[x]);
				--x; tecFileName[x] = tolower(tecFileName[x]);
				--x; tecFileName[x] = tolower(tecFileName[x]);
				if (strcmp(tecFileName + x, "map") == 0 || strcmp(tecFileName + x, "txt") == 0)
					tecFileName[x - 1] = 0;
				else
					strcpy(tecFileName, mapFileName);
			}
		}
		strcat(tecFileName, genXML ? ".xml" : ".tec");
	}

	if (mapFileName != 0) {
		// compile the mapping
		FILE*	inFile = fopen(mapFileName, "rb");
		size_t	len;
		Byte*	compiledTable;
		UInt32	compiledSize;
		TECkit_Status	status;
		char*	txt;
		
		if (inFile == 0) {
			// try adding .map
			char*	mapFileName2 = (char*)malloc(strlen(mapFileName) + 5);
			if (mapFileName2 == 0)
				return 1;
			strcpy(mapFileName2, mapFileName);
			strcat(mapFileName2, ".map");
			inFile = fopen(mapFileName2, "rb");
			free(mapFileName2);
		}
		if (inFile == 0) {
			fprintf(stderr, "unable to open mapping file %s\n", mapFileName);
			return 1;
		}
		fseek(inFile, 0, SEEK_END);
		len = ftell(inFile);
		fseek(inFile, 0, SEEK_SET);
		
		txt = (char*)malloc(len);
		if (txt == 0) {
			fprintf(stderr, "not enough memory to read mapping file\n");
			return 1;	// not enough memory
		}

		if (fread(txt, len, 1, inFile) != 1) {
			fprintf(stderr, "not enough data in mapping file\n");
			return 1;
		}
		fclose(inFile);
		
		status = TECkit_CompileOpt(txt, len, &errFunc, 0, &compiledTable, &compiledSize,
						form | (compress ? kCompilerOpts_Compress : 0) | (genXML ? kCompilerOpts_XML : 0));

		free(txt);
		
		if (status == kStatus_NoError) {
			// save the compiled mapping (or XML representation)
			FILE*	outFile;
			if (strcmp(tecFileName, "-") == 0)
				outFile = stdout;
			else {
				remove(tecFileName);
				outFile = fopen(tecFileName, "wb");
				if (outFile == 0) {
					fprintf(stderr, "unable to open output file %s\n", tecFileName);
					return 1;
				}
			}
			fwrite(compiledTable, compiledSize, 1, outFile);
			fclose(outFile);
			TECkit_DisposeCompiled(compiledTable);
		}
		else {
			fprintf(stderr, "compilation failed: status = %d\n", (int)status);
			return 1;
		}
		
		return 0;
	}

	return 1;
}
